rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -326,19 +326,30 @@ template<typename AttributeT>
 inline
 rcutils_ret_t
 _assign_partitions_to_attributes(
-  const char * const topic_name, const char * const prefix, AttributeT * attributes)
+  const char * const topic_name,
+  const char * const prefix,
+  bool avoid_ros_namespace_conventions,
+  AttributeT * attributes)
 {
   rcutils_ret_t ret = RCUTILS_RET_ERROR;
   // set topic and partitions
   rcutils_string_array_t name_tokens = rcutils_get_zero_initialized_string_array();
   name_tokens = rcutils_split_last(topic_name, '/');
   if (name_tokens.size == 1) {
-    attributes->qos.m_partition.push_back(prefix);
+    if (!avoid_ros_namespace_conventions) {
+      attributes->qos.m_partition.push_back(prefix);
+    }
     attributes->topic.topicName = name_tokens.data[0];
     ret = RCUTILS_RET_OK;
   } else if (name_tokens.size == 2) {
-    attributes->qos.m_partition.push_back(
-      (std::string(prefix) + "/" + name_tokens.data[0]).c_str());
+    if (avoid_ros_namespace_conventions) {
+      // no prefix to be used, just assign the user's namespace
+      attributes->qos.m_partition.push_back(name_tokens.data[0]);
+    } else {
+      // concat the prefix with the user's namespace
+      attributes->qos.m_partition.push_back(
+        (std::string(prefix) + "/" + name_tokens.data[0]).c_str());
+    }
```

Comment:
I updated the logic in bf1f981, but I left the two blocks in the if statement, because I think it's easier to read. But I factored the `push_back` out of the conditional.